api/src/wfl/module/covid.clj

```diff
@@ -183,17 +183,120 @@
                     (->> (jdbc/insert! tx tdr-source-table)))]
     [tdr-source-type (-> items first :id str)]))
 
+(defn ^:private find-new-rows
+  "Find new rows in TDR by querying between `last_checked` and the
+   frozen `now`."
+  [{:keys [dataset
+           dataset_table
+           table_column_name
+           last_checked] :as _source}
+   now]
+  (-> (datarepo/query-table-between
+       dataset
+       dataset_table
+       table_column_name
+       [last_checked now]
+       [:datarepo_row_id])
+      :rows
+      flatten))
+
+(defn ^:private go-create-snapshot!
+  "Create snapshot in TDR from `dataset` body, `table`
+   and `row-ids` then write job info as well as rows
+   into `source-details-name` table, `suffix` will be
+   appended to the snapshot names."
+  [suffix dataset table row-ids]
+  (let [columns     (-> (datarepo/all-columns dataset table)
+                        (->> (map :name) set)
+                        (conj "datarepo_row_id"))
+        job-id (-> (datarepo/make-snapshot-request dataset columns table row-ids)
+                   (update :name #(str % suffix))
+                   (datarepo/create-snapshot-job))]
+    job-id))
+
+(defn ^:private create-snapshots
+  "Create uniquely named snapshots in TDR with max partition size of 500,
+   using the frozen `now-obj`, from `row-ids`, return TDR job-ids."
+  [{:keys [dataset dataset_table] :as _source} now-obj row-ids]
+  (let [shards (partition-all 500 row-ids)
+        compact-now (.format now-obj (DateTimeFormatter/ofPattern "YYYYMMdd'T'HHmmss"))]
+    (doall (map-indexed (fn [idx shard]
+                          (go-create-snapshot!
+                           (format "_%s_%s" compact-now idx)
+                           dataset
+                           dataset_table
+                           shard)) shards))))
+
+(defn ^:private get-pending-tdr-jobs [{:keys [details] :as _source}]
+  (let [query "SELECT snapshot_creation_job_id
+               FROM %s
+               WHERE snapshot_creation_job_status = 'running'"]
+    (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+      (->> (format query details)
+           (jdbc/query tx)))))
+
+(defn ^:private check-tdr-job
+  "Check TDR job status for `job-id`, return a map with job-id,
+   snapshot_id and job_status if job has failed or succeeded, otherwise nil."
+  [job-id]
+  (when-let [job-metadata (datarepo/get-job-metadata-when-done job-id)]
+    (let [{:keys [id job-status] :as result} job-metadata]
+      ;; TODO: catch but not throw at (= job-status "succeeded")
+      (if (= job-status "succeeded")
```

Comment:
Yeah, I was thinking about your original "compound error` thrown at the end, and realized it requires more work here. I'd prefer we log this failure and our monitoring system (perhaps log-based stackdriver) can pickup by keywords and notify us when this happens.

When it comes to `store the state of the job so we don't always check the failed jobs`, yes, that's why I was struggling to not throw but rather log here. In this case, the `job-metadata` map already stores the failure status which will then be written to the DB. Since `get-pending-tdr-jobs` uses `WHERE snapshot_creation_job_status = 'running'`, we would always only check for running jobs not `succeeded` nor `failed`.

Reply:
Right, but are you ever writing 'failed' to the 'snapshot_creation_job_status`?